compiler/passes/scopeResolve.cpp

```diff
@@ -2042,6 +2114,73 @@ static void buildBreadthFirstModuleList(Vec<ModuleSymbol*>* modules,
   }
 }
 
+// Returns true if this module is capable of being used or traversed as part of
+// an access in the provided scope, false if the module is private and the
+// scope is not its direct parent or a sibling module when the parent of both
+// had a use of it.
+bool ModuleSymbol::isVisible(BaseAST* scope) {
+  if (hasFlag(FLAG_PRIVATE)) {
+    BaseAST* parentScope = getScope(defPoint);
+    INT_ASSERT(parentScope != NULL); // Should be true, given we found this
+    // module symbol.
+
+    if (scope == parentScope) {
+      // The scope provided is the scope in which we are defined
+      return true;
+    } else {
+      // Well, the scope is not our immediate parent.  It could be a subscope
+      // of our parent scope, though, so we need to walk up scopes until we
+      // either find our parent scope (in which case, we're visible if it "use"s
+      // us) or we run out of scope to check against (in which case we are most
+      // certainly *not* visible)
+      BaseAST* searchScope = getScope(scope);
+      bool passedModule = false;
```

Comment:
I am having a hard time with this function and so may be misreading things.  If I have any understanding of the
while loop that starts on 2138 then I think it would be more transparent if it were written as

while (searchScope != NULL && searchScope != parentScope) { 
   <simplified body> 
} 

if (searchScope == parentScope) {

}

This would make it clearer that the we are traversing upwards until we fine the parentScope or hit top.  If
we do find the parent scope then there is a well defined logic on what to return.

Separately it is intriguing to note that the "else if" part of the match on searchScope (line 2147 etc) could be executed
without running the traversal.  If I read it correctly then it is a function of the parentScope and doesn't care about
the traversal.  It might be that doing the test early would have a tiny tiny performance boost for some programs but
this is not my concern; I am only wondering if it would make the business logic more clear to a reader.